Updating logging components image defaulting pattern

ewolinetz opened the pull request:
Matches openshift_logging pattern and pattern used by other ansible components

Addresses:
https://bugzilla.redhat.com/show_bug.cgi?id=1505681

ewolinetz commented:
/kind bug

ewolinetz commented on roles/openshift_logging_elasticsearch/vars/default_images.yml:
```diff
@@ -1,3 +1,5 @@
-__openshift_logging_elasticsearch_proxy_image_prefix: "docker.io/openshift/"
+__openshift_logging_elasticsearch_image_prefix: "{{ openshift_logging_image_prefix | default('docker.io/openshift/origin-') }}"
+__openshift_logging_elasticsearch_image_version: "{{ openshift_logging_image_version | default('latest') }}"
+__openshift_logging_elasticsearch_proxy_image_prefix: "{{ openshift_logging_image_prefix | default('docker.io/openshift/origin-') | regex_replace('$(.*)/origin-', '\\1/') }}"
```
starting my regex with `$` instead of `^` ....

wozniakjan replied:
is it a good idea to use a regexp here? what if a user sets `openshift_logging_image_prefix` to `some-registry/origin-` and is more consistent with his image versioning and really means it? I think it is better not to take our inconsistencies as a precedence

wozniakjan commented:
the `openshift_logging_image_version=v3.7.0` was just for illustration purposes, but we will probably have much more consistent versioning in our downstream images than upstream. The way I see it, setting `openshift_logging_image_version` should work across all logging images the same way, override the defaults, and user can cherry pick and overwrite this per `image_component` by setting `openshift_logging_{{component_image}}_image_version`.

Btw. why do we need to set the defaults as role variables and then set as facts? I am not sure it brings anything over setting defaults as `defaults`.

ewolinetz commented:
@wozniakjan ack, addressed those concerns.

This pattern was something Scott had come up with, it allows sane defaults based on the deployment type of openshift (origin or enterprise) but it also allows you to override those defaults if you want.  There wasn't an easy way to do this otherwise as `defaults/main` will evaluate when the role is configured but we wouldn't load the default of the image version or prefix until we loaded the correct `var` file which happens in the role itself, after defaults were loaded.

wozniakjan commented:
I dare to disagree here :) I have proof that you don't have to have them as role variables and then set them as facts to be able to use defaults based on deployment type here: https://github.com/openshift/openshift-ansible/pull/5908 and we have had that in 1.5, 3.6 and master for some time without an issue.

ewolinetz commented:
@wozniakjan I understand that its possible. 

To me, this makes it easier to distinguish between enterprise and origin OOTB and provide sane defaults, keeps the roles from being dependent on `openshift_logging` to provide sane defaults, and follows a pattern that is being used in other openshift-ansible roles. Previously we only defaulted within these roles for origin installations and relied on `openshift_logging` to give us defaults -- that seemed like a fragile design.

ewolinetz commented:
/retest

ewolinetz commented:
test is failing because we dont have an `es_pod` here https://github.com/openshift/origin-aggregated-logging/blob/master/test/zzz-correct-index-names.sh#L29

ewolinetz commented:
/retest

richm commented:
> test is failing because we dont have an es_pod here https://github.com/openshift/origin-aggregated-logging/blob/master/test/zzz-correct-index-names.sh#L29

How can `es_pod` be missing or empty here, unless elasticsearch is not running?

wozniakjan commented:
es_pod is missing because the default values are not correct and it tries to pull nonexistent image:tag.

@ewolinetz this also still breaks on basic scenarios

default values and no changes in the inventory:
```
docker.io/openshift/oauth-proxy:latest                 <-- should be docker.io/openshift/oauth-proxy:v1.0.0
docker.io/openshift/origin-logging-auth-proxy:latest
docker.io/openshift/origin-logging-curator:latest
docker.io/openshift/origin-logging-elasticsearch:latest
docker.io/openshift/origin-logging-fluentd:latest
docker.io/openshift/origin-logging-kibana:latest
```

setting `openshift_logging_image_prefix=local-reg:5000/openshift/origin-`:
```
local-reg:5000/openshift/oauth-proxy:latest             <-- should be local-reg:5000/openshift/origin-oauth-proxy:v1.0.0 so this is wrong for both tag and prefix
local-reg:5000/openshift/origin-logging-auth-proxy:latest
local-reg:5000/openshift/origin-logging-curator:latest
local-reg:5000/openshift/origin-logging-elasticsearch:latest
local-reg:5000/openshift/origin-logging-fluentd:latest
local-reg:5000/openshift/origin-logging-kibana:latest
```

I really think we shouldn't spend more time here unless we find something broken with #5908 because this aims to provide exactly the same fix as https://github.com/openshift/openshift-ansible/pull/5908, but #5908 works for all known scenarios.

wozniakjan commented:
from the `https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_openshift-ansible/6037/test_pull_request_openshift_ansible_logging/3037/artifacts/scripts/entrypoint/artifacts/events.txt`
it looks like that is the case, instead of pulling `openshift/oauth-proxy:v1.0.0` it goes for `openshift/oauth-proxy:latest`
```
logging                26m        31m         20        logging-es-data-master-qv6j3a96-1-94c6h   Pod                     spec.containers{proxy}                Normal    BackOff                   kubelet, ip-172-18-3-200.ec2.internal         Back-off pulling image "openshift/oauth-proxy:latest"
logging                30m        31m         4         logging-es-data-master-qv6j3a96-1-94c6h   Pod                     spec.containers{proxy}                Warning   Failed                    kubelet, ip-172-18-3-200.ec2.internal         Error: ImagePullBackOff
```

ewolinetz commented:
currently just failing from logging flake (I believe its the fluentd related one)

ewolinetz commented:
/retest

ewolinetz commented:
@wozniakjan this is currently written the way it is (where for `origin` installations we make minor exceptions/updates for the ES proxy) due to the fact that there isn't a matching tag that is in step with the other logging components.

To me, that should not be a blocker for the PR as it is to resolve a bug where we do not pass the prefix on for origin installations which is what this PR addresses -- we can remove the exceptions later on as tech-debt once the tags are available. This also unifies the image version pattern with other `openshift-ansible` components such as `openshift_logging`, `template_service_broker`, and `ansible_service_broker`. Otherwise, this functionally is the same as #5908

wozniakjan commented:
nack, doesn't work for at least one common scenario

setting `openshift_logging_image_version=v3.7.0` should really work for all images:
```
docker.io/openshift/oauth-proxy:v1.0.0             <-- should be docker.io/openshift/oauth-proxy:v3.7.0
docker.io/openshift/origin-logging-auth-proxy:v3.7.0
docker.io/openshift/origin-logging-curator:v3.7.0
docker.io/openshift/origin-logging-elasticsearch:v3.7.0
docker.io/openshift/origin-logging-fluentd:v3.7.0
docker.io/openshift/origin-logging-kibana:v3.7.0
```

and I don't think the fact it still breaks on `setting openshift_logging_image_prefix=[random_registry]/openshift/origin-` is a techdebt, but it will be a bz filled by our QE as soon as this merges

ewolinetz commented:
/test logging

ewolinetz commented:
doesn't look like it picked up the new inventory vars...
/retest

ewolinetz commented:
I was told they'd be updated now... 
/retest

richm commented:
I was told there would be cake . . .
/retest

wozniakjan commented on roles/openshift_logging_elasticsearch/vars/default_images.yml:
```diff
@@ -1,5 +1,5 @@
 __openshift_logging_elasticsearch_image_prefix: "{{ openshift_logging_image_prefix | default('docker.io/openshift/origin-') }}"
 __openshift_logging_elasticsearch_image_version: "{{ openshift_logging_image_version | default('latest') }}"
-__openshift_logging_elasticsearch_proxy_image_prefix: "{{ openshift_logging_image_prefix | default('docker.io/openshift/origin-') | regex_replace('^(.*)/origin-$', '\\1/') }}"
+__openshift_logging_elasticsearch_proxy_image_prefix: "{{ openshift_logging_image_prefix | default('docker.io/openshift/') }}"
```
this will not work, you set `openshift_logging_image_prefix` as a fact here:
https://github.com/ewolinetz/openshift-ansible/blob/6575595891f71e7d9b9ac929b9ba58117a65e17b/roles/openshift_logging/tasks/main.yaml#L16-L17

therefore it is never null and usage of `default()` makes no sense here

ewolinetz replied:
This is so that if the role is used standalone we will still have a default.

ewolinetz replied:
@wozniakjan ah ok, I see what you're saying here... so if we remove the fact setting in `openshift_logging` then we can properly fall back on the defaults for the roles. I was missing that point before...

ewolinetz commented:
/retest

ewolinetz commented:
/retest

ewolinetz commented:
/retest

jcantrill commented:
/lgtm

ewolinetz commented:
/cherrypick release-3.7
